src/core/storage/client.ts

```diff
@@ -0,0 +1,168 @@
+   */
+  async initialize(): Promise<void> {
+    if (this.initialized) {
+      return;
+    }
+
+    if (!this.config.privateKey) {
+      throw new Error('Private key is required');
+    }
+
+    if (!this.config.delegation) {
+      throw new Error('Delegation is required');
+    }
+
+    try {
+      const principal = Signer.parse(this.config.privateKey);
+      const store = new StoreMemory();
+      this.storage = await Storage.create({ principal, store });
+
+      const delegationProof = await parseDelegation(this.config.delegation);
+      const space = await this.storage.addSpace(delegationProof);
+      await this.storage.setCurrentSpace(space.did());
+
+      this.initialized = true;
+    } catch (error: unknown) {
+      const message = error instanceof Error ? error.message : 'Unknown error';
+      throw new Error(`Failed to initialize storage client: ${message}`);
+    }
+  }
+
+  /**
+   * Get the storage client
+   * @returns The storage client
+   */
+  getStorage(): Storage.Client | null {
+    return this.storage;
+  }
+
+  /**
+   * Check if the client is connected and ready
+   */
+  isConnected(): boolean {
+    return this.initialized;
+  }
+
+  /**
+   * Get the storage config
+   * @returns The storage config
+   */
+  getConfig(): StorageConfig {
+    return this.config;
+  }
+
+  /**
+   * Get the configured gateway URL
+   * @returns The configured gateway URL
+   * @throws Error if gateway URL is not set
+   */
+  getGatewayUrl(): string {
+    if (!this.config.gatewayUrl?.trim()) {
+      throw new Error('Gateway URL is not set');
+    }
+    return this.config.gatewayUrl;
+  }
+
+  /**
+   * Upload files to Storacha network
+   * The Storage Client needs to be initialized to upload files.
+   * 
+   * @param files - Array of files to upload
+   * @param options - Upload options
+   * @returns The uploaded files' Content ID and URL
+   */
+  async uploadFiles(files: UploadFile[], options: UploadOptions = {}): Promise<UploadResult> {
+    if (!this.initialized || !this.storage) {
+      throw new Error('Client not initialized');
+    }
+
+    try {
+      if (options.signal?.aborted) {
+        throw new Error('Upload aborted');
+      }
+
+      const fileObjects = files.map(file => {
+        const buffer = Buffer.from(file.content, 'base64');
+        return new File([buffer], file.name, {
+          type: file.type
+        });
+      });
+
+      const root = await this.storage.uploadDirectory(fileObjects, {
+        // If publishToIPFS is false, we don't provide a pieceHasher, so the content is not pinned to IPFS
+        ...(options.publishToIPFS === false ? { pieceHasher: undefined } : {}),
+        retries: options.retries ?? 3,
+        signal: options.signal
+      });
+
+      return {
+        root: root.toString(),
+        rootURL: new URL(`/ipfs/${root.toString()}`, this.getGatewayUrl()).toString(),
+        files: files.map(file => ({
```

Comment:
You can use `onDirectoryEntryLink` to collect and return the CIDs for each of these files.

https://github.com/storacha/upload-service/blob/576fc943e919e7a7a583551c7ce378b3d2cea0b6/packages/upload-client/src/types.ts#L337-L340